### PR TITLE
chore: update ipfs (now with typing) dependencies

### PR DESCRIPTION
**Pull request by achingbrain**
This follows on from #1194 with a few extra bits and pieces (can target that branch for merging if it's preferable):

- Updates types
- Removes [multiformats](https://www.npmjs.com/package/multiformats) in favour of [blockcodec-to-ipld-format](http://npmjs.com/package/blockcodec-to-ipld-format) as it outputs the correct types
- Swaps `ipfs` for `ipfs-core` - no need to pull down cli deps etc if they're not being used
- Updates `ipld-dag-cbor` to the latest version
- Depends on `next` for `ipfs-*` to get the latest rc

### CODE REVIEW

**Comment by oed** [comment]
Thanks @achingbrain looking forward to having native ipfs typings! 

Will let others review this PR in depth.

> Depends on next for ipfs-* to get the latest rc

I assume this PR is a draft until new IPFS versions are release? :)

**Comment by achingbrain** [comment]
Yes!  Sorry, I should have opened it as a draft..

**Comment by achingbrain** [comment]
The following commands in this repo now work for me:

```console
$ npm install
$ npm run build
$ npm test
```

Please let me know if there's anything else I should be running..

**Comment by oed** [comment]
What node version are you on @achingbrain ?

What errors are you seeing when running npm install?

**Comment by achingbrain** [comment]
> What node version are you on @achingbrain ?

14.  I tried 15 but saw errors related to peer dependencies caused by the change in how npm@7 handles them.

> What errors are you seeing when running npm install?

None, like I said, the commands above work without error 😄

**Comment by oed** [comment]
Oh, haha!

I misread 😹

**Comment by oed** [comment]
You should also run `npm run lint`.

**Comment by achingbrain** [comment]
There's one `@ts-ignore` that needs removing, just needs https://github.com/ipfs/js-ipfs/pull/3660 landing.

**Comment by achingbrain** [comment]
That's been merged & released as an rc, linting is happy now.

**Review by stbrody — COMMENTED** [review]
A few very minor comments but overall this looks really great.  I'm excited to get this merged in.

I'd still like @ukstv to take a look as well to double-check me, especially around the serialization/deserialization stuff in pubsub.

File: packages/core/src/__tests__/ipfs-util.ts [hunk 1]
```diff
@@ -27,6 +24,7 @@ export async function createIPFS(overrideConfig: Record<string, unknown> = {}):
   };
 
   const config = { ...defaultConfig, ...overrideConfig };
+  // TODO: @rvagg dag-jose ipld format?
```
[the remark this answers is not in the document]
Comment (oed) [reply packages/core/src/__tests__/ipfs-util.ts:27]: No, this is pending a PR to https://github.com/ceramicnetwork/js-dag-jose I think.

**Comment by achingbrain** [comment]
@stbrody I think I've removed all the outstanding ts-ignores now, as well as the double-serialization of the pubsub message.

I'm going to flesh out the test suite for `blockcodec-to-ipld-format` and publish a `v1` but other than that please feel free to point out any bits of the API that look odd to you or other points of friction.

**Comment by achingbrain** [comment]
Everything's been released so this should be good to go, pending anything further review uncovers.

File: packages/ipfs-daemon/src/ipfs-daemon.ts [hunk 2]
```diff
@@ -148,9 +142,8 @@ export class IpfsDaemon {
                     ...configuration.ipfsEnableApi && {API: `/ip4/${configuration.tcpHost}/tcp/${configuration.ipfsApiPort}`},
                     ...configuration.ipfsEnableGateway && {Gateway: `/ip4/${configuration.tcpHost}/tcp/${configuration.ipfsGatewayPort}`}
                     ,
+                    Announce: configuration.announceAddressList,
```
Comment (oed) [inline packages/ipfs-daemon/src/ipfs-daemon.ts:145]: Why is this passed to both `config` and `libp2p.config`?

**Review by oed — APPROVED** [review]
lgtm :)

**Comment by oed** [comment]
We should make sure that all of these work:
```
$ npm install
$ npm run build
$ npm test
$ npm run lint
$ npm run docs
```

File: packages/ipfs-daemon/src/ipfs-daemon.ts (hunk 2 above)
Comment (achingbrain) [reply packages/ipfs-daemon/src/ipfs-daemon.ts:145]: I was actually wondering the same thing, it shouldn't be necessary to pass the list in both places.

File: packages/ipfs-topology/src/ipfs-topology.ts [hunk 3]
```diff
@@ -109,19 +110,17 @@ export class IpfsTopology {
...
   private async _forceBootstrapConnection(
     ipfs: IpfsApi,
-    bootstrapList: string[]
+    bootstrapList: Multiaddr[]
   ): Promise<void> {
     // Don't want to swarm connect to ourself
     const myPeerId = (await ipfs.id()).id
     const filteredBootstrapList = bootstrapList.filter((addr) => {
-      return !addr.endsWith(myPeerId)
+      return !addr.getPeerId()?.endsWith(myPeerId)
```
Comment (stbrody) [inline packages/ipfs-topology/src/ipfs-topology.ts:118]: @achingbrain, I pulled down this PR and updated all the ipfs deps in the various `package.json` files to use `next`.  The tests all passed, but running `ceramic daemon` in a terminal causes this line to throw the error:
```
Connecting to peers found in 'https://raw.githubusercontent.com/ceramicnetwork/peerlist/main/testnet-clay.json'
(node:126315) UnhandledPromiseRejectionWarning: TypeError: addr.getPeerId is not a function
    at /home/spencer/js-ceramic4/packages/ipfs-topology/lib/ipfs-topology.js:105:33
    at Array.filter (<anonymous>)
    at IpfsTopology._forceBootstrapConnection (/home/spencer/js-ceramic4/packages/ipfs-topology/lib/ipfs-topology.js:103:53)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async IpfsTopology.forceConnection (/home/spencer/js-ceramic4/packages/ipfs-topology/lib/ipfs-topology.js:78:9)
    at async IpfsTopology.start (/home/spencer/js-ceramic4/packages/ipfs-topology/lib/ipfs-topology.js:81:9)
    at async Ceramic._init (/home/spencer/js-ceramic4/packages/core/lib/ceramic.js:257:13)
    at async Function.create (/home/spencer/js-ceramic4/packages/cli/lib/ceramic-daemon.js:124:9)
    at async Command.<anonymous> (/home/spencer/js-ceramic4/packages/cli/lib/bin/ceramic.js:34:5)

```

Comment (oed) [reply packages/ipfs-topology/src/ipfs-topology.ts:118]: The multiaddress in the file from there are in string and needs to be converted into a Multiaddress instance.

File: packages/ipfs-topology/src/ipfs-topology.ts [hunk 4]
```diff
@@ -104,24 +105,22 @@ export class IpfsTopology {
     }
     this.logger.imp(`Connecting to peers found in '${url}'`);
     const list = await fetchJson(url);
-    return list || [];
+    return list.map(peer => new Multiaddr(peer)) || [];
```
Comment (stbrody) [inline packages/ipfs-topology/src/ipfs-topology.ts:108]: This change fixes the ability to run the daemon from the CLI

File: packages/ipfs-topology/src/ipfs-topology.ts (hunk 3 above)
Comment (stbrody) [reply packages/ipfs-topology/src/ipfs-topology.ts:118]: Yep that was it.  Pushed a fix to this PR branch, seems to be working now.  I was able to run a node on the clay testnet, create and update a stream, and then load that stream from the clay gateway

File: packages/ipfs-topology/src/ipfs-topology.ts (hunk 4 above)
Comment (oed) [reply packages/ipfs-topology/src/ipfs-topology.ts:108]: This will break if `list` is undefined, if it's never undefined then we don't need `|| []`.

Comment (stbrody) [inline packages/ipfs-topology/src/ipfs-topology.ts:108]:
```suggestion
    return list?.map(peer => new Multiaddr(peer)) || [];
```

**Comment by stbrody** [comment]
> We should make sure that all of these work:
> 
> ```
> $ npm install
> $ npm run build
> $ npm test
> $ npm run lint
> $ npm run docs
> ```

Done.  All working

**Pull request merged**